### PR TITLE
add colour hint to resources with missing files

### PR DESCRIPTION
**Pull request by blurymind**
this addresses the rest of #810 - the resource list in the resource manager will now indicate with a red colour if a file can't be found

This is of course in line with the way the themes work, so it can be customized via the theme's index file

errorhinting-gdresourcemanager
errorhinting-gdresourcemanager

### CODE REVIEW

**Comment by 4ian** [comment]
Awesome :) Let me check this this week-end. After a super quick look, I think we'll move the computation of the errored files outside of the render (in a React lifecycle method) to avoid querying the file system at every render, which can be quickly expensive.
The design by itself looks great :)

**Comment by blurymind** [comment]
There are two things to consider:
- When do we run a check on all files to see if they are still there and havent been moved?
when loading the project, when triggered manually, on playtesting?
- How can we transfer this functionality over to other places where resources are listed in a generic and inexpensive way?

**Comment by blurymind** [comment]
how about we do the check only when the component is mounted?

**Comment by blurymind** [comment]
I wonder if react has something like a virtual scroll element
https://www.youtube.com/watch?v=UtD41bn6kJ0

perhaps it could help performance in rendering very very big lists?

**Comment by 4ian** [comment]
Lists are already virtualized ;) 

https://github.com/4ian/GDevelop/blob/238b6a2a092f38e0f572a3872115fd1708caa5b2/newIDE/app/src/UI/SortableVirtualizedItemList/index.js#L3

> how about we do the missing path check only when the component is mounted? Should be much cheaper that way.

Yes let's do it only when the component is mounted.

Ideally, we would watch the file system, but it's too much effort for something that is not worth it I think. This is because most users will open the resource editor from time to time, after a while working on the game, and will see at this moment missing files. It will be enough.

**Comment by blurymind** [comment]
react is awesome! Btw I already made it only trigger on mount

File: newIDE/app/src/UI/SortableVirtualizedItemList/index.js [hunk 1]
```diff
@@ -95,6 +97,7 @@ class ItemsList extends React.Component<ItemsListProps, *> {
               selected={item === selectedItem}
               onItemSelected={this.props.onItemSelected}
               buildMenuTemplate={this.props.buildMenuTemplate}
+              hasErrors={errorItemsList.includes(item.getName())}
```
Comment (4ian) [inline newIDE/app/src/UI/SortableVirtualizedItemList/index.js]: Finally, here to `hasErrors={!!erroredItems[item.getName()]}`

Why? Because "includes" will do a search in the table of erroredItems.
Let's say we have 1000 resources, 50 items displayed on the screen, and 200 resources that are errored.

Searching in an array need to traverse the whole array, in the worst case. If we call N the length of the array, we say that the complexity is "N".
Here, our array of errors is 200 items long. So we're going to search in this array for all the 50 items display on screen. We're going to do 50*200 = 10 000 operations (in the worst case).

If instead we have an object, we can improve this. Why?
Because objects are stored internally as a "hashmap". Hashmaps are super great, because given a "key", they given you the value in constant time. (we say the complexity is 1)

Let's say we have 1000 resources, 50 items displayed on the screen, and 200 resources that are errored.
Now, our errors are stored in an object.
For all the 50 items, we're going to get the value for the key being the name of the resource. We're going to do 50*1 = 50 operations. Way faster! And if you have a lot of errors, it does not even matter, the number of operations stay constant, which is great.

Creating the hash map at the component mounting need to traverse the array of resources, which is done only once and very quick. The cost of creating this object ("hashmap") of errored items is so negligible and a great deal because it optimize a lot every other rendering :)

Comment (4ian) [reply newIDE/app/src/UI/SortableVirtualizedItemList/index.js]: If you're interested, it's a bit technical but search for "complexity of data structure" : http://blog.benoitvallon.com/data-structures-in-javascript/data-structures-in-javascript/
This looks like a gentle introduction to this field: https://medium.freecodecamp.org/time-is-complex-but-priceless-f0abd015063c

These are things that are a bit mathematical and more in the computer science field. You might not be interested, but still worth knowing a few fact.
Basically: searching in a big array = slow and searching/setting a value in an object = quick. The advantage of an array is that it's ordered... but here we don't care about the ordering of the errors ;) We just want to get them quick!

Comment (blurymind) [reply newIDE/app/src/UI/SortableVirtualizedItemList/index.js]: @4ian this is gold! Thank you for taking the time to explain 😄 I am making the amendments now

Comment (blurymind) [reply newIDE/app/src/UI/SortableVirtualizedItemList/index.js]: fantastic articles too. Thank you again for sharing this. I committed the changes now

**Comment by 4ian** [comment]
Tests are failing because of a update to libGD.js (delete it in app/public and relaunch yarn start/npm start by the way). I've fixed it on master, you can merge/rebase on master to have the tests passing.

You can run `yarn flow` on your computer to get flow checking (don't forget to run prettier too as usual)

File: newIDE/app/src/ResourcesList/index.js [hunk 2]
```diff
@@ -62,6 +63,7 @@ export default class ResourcesList extends React.Component<Props, State> {
   state: State = {
     renamedResource: null,
     searchText: '',
+    resourcesWithMissingPath: {'': false},
```
[the remark this answers is not in the document]
Comment (blurymind) [reply newIDE/app/src/ResourcesList/index.js]: ah sorry, my brain is scrambled when coding late. xD I have just noticed it now

**Comment by blurymind** [comment]
running flow locally strangely takes longer than having the server do the test, but  obviously its not good to spam the commit history will attempts at flow fixes. Sorry about that :)

**Comment by 4ian** [comment]
Thanks for merging master :) 
Run `yarn format` to ensure that all files are properly formatted (from now, Semaphore CI and Travis CI will warn and error if there is a not formatted file - to ensure that we keep a always 100% formatted codebase) and if there is something changed, you can push the changes.

**Comment by 4ian** [comment]
@blurymind **Don't push expression parser2 here!!!** Can you undo this?

**Comment by blurymind** [comment]
oops sorry, I merged it by mistake. I meant to merge master

**Comment by 4ian** [comment]
Cool, now you can run `yarn` and `yarn format` :)

**Comment by blurymind** [comment]
@4ian all done now :)
Btw yarn format is fantastic! I am really liking this

**Comment by blurymind** [comment]
You might have to add a note to the instructions about using it. Also I had to install yarn to do it. Prior to that I have been using only npm, so I was wondering if we can run the format with npm

**Comment by 4ian** [comment]
Yes, `npm run format`. Format is just a script that I put in `package.json` in newIDE/app that run prettier ;) I'll add something in the README.

**Comment by 4ian** [comment]
I've added the command in https://github.com/4ian/GDevelop/blob/master/newIDE/README.md#3-how-to-contribute- and also in the section about tests ;)

**Comment by 4ian** [comment]
Looks good, I've merged this! Thanks for this addition 👍

**Comment by blurymind** [comment]
thank you :)
